server/docs/design/bidi-producer-consumers-streaming.md

```diff
@@ -0,0 +1,339 @@
+
+**subscribe** - Consumers calling the `StreamSource` gRPC service must be affiliated or subscribed with a producer to 
+receive a live stream of BlockItems from the `hedera-block-node`.
+
+**unsubscribe** - Consumers terminating their connection with the `StreamSource` gRPC service must be unaffiliated or 
+unsubscribed from a producer so that internal objects can be cleaned up and resources released.
+
+---
+
+### Block Node gRPC Streaming Services API
+
+The following protobuf definition outlines the gRPC services and messages used to stream BlockItems between a producer,
+consumers and the `hedera-block-node`.  The `BlockStreamGrpc` service definition provides 2 bidirectional streaming
+methods: `StreamSink` and `StreamSource`.  Aside from the gRPC service and method names, all the types will be replaced 
+by those outlined in [hedera-protobufs](https://github.com/hashgraph/hedera-protobufs/pull/342/files).
+
+```protobuf
+
+/**
+  * The BlockStreamGrpc service definition provides 2 bidirectional streaming methods for
+  * exchanging BlockItems with the Block Node server.
+  *
+  * A producer (e.g. Consensus Node) can use the StreamSink method to stream BlockItems to the
+  * Block Node server.  The Block Node server will respond with a BlockResponse message for
+  * each BlockItem received.
+  *
+  * A consumer (e.g. Mirror Node) can use the StreamSource method to request a stream of
+  * BlockItems from the server.  The consumer is expected to respond with a BlockResponse message
+  * with the id of each BlockItem received.
+ */
+service BlockStreamGrpc {
+  /**
+  * StreamSink is a bidirectional streaming method that allows a producer to stream BlockItems 
+  * to the Block Node server.  The server will respond with a BlockResponse message for each
+  * BlockItem received.
+  */
+  rpc StreamSink (stream BlockItem) returns (stream BlockItemResponse) {}
+
+  /**
+    * StreamSource is a bidirectional streaming method that allows a consumer to request a
+    * stream of BlockItems from the server.  The consumer is expected to respond with a BlockResponse
+    * message with the id of each BlockItem received.
+    */
+  rpc StreamSource (stream BlockItemResponse) returns (stream BlockItem) {}
+}
+
+/**
+ * A BlockItem is a simple message that contains an id and a value.
+ * This specification is a simple example meant to expedite development.
+ * It will be replaced with a PBJ implementation in the future.
+ */
+message BlockItem {
+  /**
+   * The id of the block.  Each block id should be unique.
+   */
+  int64 id = 1;
+
+  /**
+   * The value of the block.  The value can be any string.
+   */
+  string value = 2;
+}
+
+/**
+ * A BlockItemResponse is a simple message that contains an id.
+ * The BlockItemResponse is meant to confirm the receipt of a BlockItem.
+ * A future use case may expand on this type to communicate a failure
+ * condition where the BlockItem needs to be resent, etc.
+ */
+message BlockItemResponse {
```

Comment:
I am not sure we need a response to every item. One per block is plenty. It seems a waste of bandwidth and effort to send one per item. If we have 100,000 items a second then that is an extra 100,000 objects that have to be created serialized, sent and garbage collected. 100,000 items is not a crazy number is is only 10 per user transaction which is not far fetched.

Reply:
Also the design intent of the feedback was to solve a problem for the consensus node. It has to keep a copy of pre-consensus event stream for events till it is sure those events have been persisted by a block node on disk.  So the response back should be once per block and after it is 100% on disk

Reply:
I can adjust the doc and message type to be `BlockResponse` to communicate that the consumer should only send 1 response per Block received.  Is this a type we need to add to the `hedera-protobufs` or does something like this already exist?

Reply:
@jsync-swirlds I think this is matching up with [PublishStreamResponse](https://github.com/hashgraph/hedera-protobufs/blob/continue-block-node/block/block_service.proto#L110) do we need to rethink this?

Reply:
I believe the current service definition accomplishes what Jasper is asking for.  The timing of _when_ the response is sent is more in the protocol realm, and if we want to start by only acknowledging once per block, that's fine.

Reply:
Good point...Joseph and I were just discussing the overlap with the API services defined in [hedera-protobufs](https://github.com/hashgraph/hedera-protobufs/blob/25783427575ded59d06d6bf1ed253fd24ef3c437/block/block_service.proto#L675).  I'll rename these to align with those names and types.

Reply:
I removed the API definition and replaced it with a link to the `hedera-protobufs` repo and the branch where to find the correct service definitions and types.